src/ot/normalize.rs

```diff
@@ -0,0 +1,500 @@
+    if ctx.buffer.cur(0).general_category() == GeneralCategory::SpaceSeparator {
+        if let Some(space_type) = u.space_fallback() {
+            if let Some(space_glyph) = ctx.font.glyph_index(' ' as u32) {
+                ctx.buffer.cur_mut(0).set_space_fallback(space_type);
+                next_char(ctx.buffer, space_glyph.0 as u32);
+                ctx.buffer.scratch_flags |= BufferScratchFlags::HAS_SPACE_FALLBACK;
+                return;
+            }
+        }
+    }
+
+    // U+2011 is the only sensible character that is a no-break version of another character
+    // and not a space.  The space ones are handled already.  Handle this lone one.
+    if u as u32 == 0x2011 {
+        if let Some(other_glyph) = ctx.font.glyph_index(0x2010) {
+            next_char(ctx.buffer, other_glyph.0 as u32);
+            return;
+        }
+    }
+
+    // Insert a .notdef glyph if decomposition failed.
+    next_char(ctx.buffer, 0);
+}
+
+/// Returns 0 if didn't decompose, number of resulting characters otherwise.
+fn decompose(ctx: &mut ShapeNormalizeContext, shortest: bool, ab: u32) -> u32 {
+    let (a, b) = match ctx.decompose(ab) {
+        Some(decomposed) => decomposed,
+        _ => return 0,
+    };
+
+    let a_glyph = ctx.font.glyph_index(a);
+    let b_glyph = if b != 0 {
+        match ctx.font.glyph_index(b) {
+            Some(glyph_id) => Some(glyph_id),
+            None => return 0,
+        }
+    } else {
+        None
+    };
+
+    if !shortest || a_glyph.is_none() {
+        let ret = decompose(ctx, shortest, a);
+        if ret != 0 {
+            if let Some(b_glyph) = b_glyph {
+                output_char(ctx.buffer, b, b_glyph.0 as u32);
+                return ret + 1;
+            }
+            return ret;
+        }
+    }
+
+    if let Some(a_glyph) = a_glyph {
+        // Output a and b.
+        output_char(ctx.buffer, a, a_glyph.0 as u32);
+        if let Some(b_glyph) = b_glyph {
+            output_char(ctx.buffer, b, b_glyph.0 as u32);
+            return 2;
+        }
+        return 1;
+    }
+
+    0
+}
+
+extern "C" fn decompose_unicode(
+    _: *const ffi::rb_ot_shape_normalize_context_t,
+    ab: ffi::rb_codepoint_t,
+    a: *mut ffi::rb_codepoint_t,
+    b: *mut ffi::rb_codepoint_t,
+) -> ffi::rb_bool_t {
+    crate::unicode::rb_ucd_decompose(ab, a, b)
+}
+
+extern "C" fn compose_unicode(
+    _: *const ffi::rb_ot_shape_normalize_context_t,
+    a: ffi::rb_codepoint_t,
+    b: ffi::rb_codepoint_t,
+    ab: *mut ffi::rb_codepoint_t,
+) -> ffi::rb_bool_t {
+    crate::unicode::rb_ucd_compose(a, b, ab)
+}
+
+fn compare_combining_class(pa: &GlyphInfo, pb: &GlyphInfo) -> bool {
+    pa.modified_combining_class() > pb.modified_combining_class()
+}
+
+fn set_glyph(info: &mut GlyphInfo, font: &Font) {
+    if let Some(glyph_id) = font.glyph_index(info.codepoint) {
+        *info.glyph_index() = glyph_id.0 as u32;
+    }
+}
+
+fn output_char(buffer: &mut Buffer, unichar: u32, glyph: u32) {
```

Comment:
Let's make it as `Buffer` methods.

Reply:
Yes, I also thought about that, but went with the minimally invasive solution first.